### ISSUE TITLE
[Bug] Dark Reader is enabled even if a website dark theme is detected

### ISSUE THREAD

**Issue by lumynou5** [darkreader/darkreader#11990]
### Prerequisites

- [X] I [searched for any existing report](https://github.com/darkreader/darkreader/issues?q=is%3Aissue) about this bug to avoid opening a duplicate.
- [X] I can reproduce this bug in a new, unmodified web browser profile with Dark Reader installed as the only extension.
- [X] I understand I need to use the [Broken Website Report template](https://github.com/darkreader/darkreader/issues/new?assignees=&labels=Broken+Website&template=broken-website-report.yml&title=%5BBroken+Website%5D+Replace+with+title) if this bug I am reporting occurs on a single website.

### Bug Description

This is not dark themes not detected, which is a known issue and what #9016 megathread is for.

The dark reader is enabled even if a dark theme is detected (the pannel shows "Dark theme detected").
Currently I found it happens on [GitHub](https://github.com/) and [YouTube](https://www.youtube.com/), but it seems not limited on these websites.

### Website Address

https://github.com/

### Steps To Reproduce

I cannot reproduce constantly.

1. Make sure Dark Theme Detect is on.
2. Visit any website with a detect-able dark theme.
3. Find out that the extension dark theme might be enabled while the website dark theme is detected.

### Expected Behavior

The dark reader is disabled automatically when the website dark theme is detected.

### Actual Behavior

The dark reader is enabled sometimes when the website dark theme is detected.

### Screenshots

It shows "dark theme detected" but the extension dark theme is actually enabled:
![Bug](https://github.com/darkreader/darkreader/assets/84374077/61211dd2-d954-4f24-9822-efdc23bfb4f2)

And it should be:
![Expected](https://github.com/darkreader/darkreader/assets/84374077/4560e38f-321c-4e82-b0bf-a465edcd0bbf)

### Operating System

Windows 10

### Web Browser name and version

Firefox 120.0.1

### Dark Reader version

4.9.73

**Comment by iamandrewluca** [comment]
@lumynou5 what are your GitHub appearance settings?

The detector for github.com right now is `[data-color-mode="dark"]` but when you change the GitHub appearance settings to be the System, `data-color-mode` becomes `auto` which does not match `[data-color-mode="dark"]`


https://github.com/darkreader/darkreader/blob/e2557931a23a6a77e651c26c56eb6c8eb2bab09f/src/config/detector-hints.config#L197-L203

**Comment by Myshor** [comment]
I merged that fix but it will be possible to test when new Dark Reader will go out so not closing the issue yet.

**Comment by lumynou5** [comment]
> @lumynou5 what are your GitHub appearance settings?
> 
> The detector for github.com right now is `[data-color-mode="dark"]` but when you change the GitHub appearance settings to be the System, `data-color-mode` becomes `auto` which does not match `[data-color-mode="dark"]`

I use Dark. I guess that if the reason is the System option cannot be detected, it should be un-detect-able always instead of showing detected and working as undetected.

**Comment by lumynou5** [comment]
Instance:
![screenshot](https://github.com/darkreader/darkreader/assets/84374077/3f80c22e-d84e-44fd-9401-f695e5384d8b)
The extension says "dark theme detected", and the detector hint (`html[dark]`) is indeed matched, but the dark theme is enabled.

**Comment by Myshor** [comment]
From that screenshot I see that dark theme from Dark Reader is disabled for youtube.
X on the left side of URL tells me that.
![obraz](https://github.com/darkreader/darkreader/assets/56877029/e2223feb-f52b-470a-8610-d3ac188c1d7a)

If Dark Reader Dark mode would be used on the site it would show V instead of X.
Like below: (I don't use Dark Theme detection option so even Dark Theme sites are default processed by Dark Reader.)
![obraz](https://github.com/darkreader/darkreader/assets/56877029/68d5365d-d6ad-457e-8bab-5b33e04f9eea)

Now I see that in the first post there was X on the left of Github URL so Github was processing was disabled too... It was hard to cacth is small version of picture and big one don't load to me.

**Comment by Kevnkkm** [comment]
> From that screenshot I see that dark theme from Dark Reader is disabled for youtube. X on the left side of URL tells me that. 

> If Dark Reader Dark mode would be used on the site it would show V instead of X. Like below: (I don't use Dark Theme detection option so even Dark Theme sites are default processed by Dark Reader.)

Even with the X indicating that it is not enabled for a website, it is still enabled. You have to click on it to "enable" it (does nothing) and click again to _actually_ disable it. On the YouTube screenshot, you can see that the colors don't match the normal dark YouTube theme (most noticeable is the search bar) because DarkReader is changing the styles.

**Comment by Steamvoker** [comment]
I have the same problem regarding YouTube. Open a page in with youtube's dark theme enabled; the extension says it has detected dark theme, but still enabled it's own dark theme over it. I toggle it on (when toggling on, nothing happens because it's already on) and then off again, it goes back to normal until I close and open than page again.

**Comment by ajhall** [comment]
I've been having this problem on YouTube as well. I ran a Git bisect and narrowed it down to this commit: https://github.com/darkreader/darkreader/commit/999201ee1d28e8b033119201d19b2eeb5df7880e

The current latest commit on the main branch (826148901b74361dbe16b4f490495063ad703f82) still has this problem, so it hasn't been fixed in any recent commits.

**Comment by SrS2225a** [comment]
I been experiencing the same issue for another website called stedi, an edi reference site. It only happens when switching between edi versions for an release and refreshing the page, then dark theme will be enabled for that site. You may need to try a few times though as it is a hit and miss whenever it does detect dark mode for stedi

**Comment by ashutoshbsathe** [comment]
For me it's not ``sometimes'', it's always! Dark mode plugin is activated on GitHub and YouTube even when the plugin successfully detects dark mode. I'm on Firefox 125 on Arch Linux

**Comment by paulosouuza** [comment]
I use Firefox 126 on Windows 11 and I too have this bug. It's very reproducible every time a new tab/window on both this websites its opened. But if you refresh the page or go to another link/session of the website STILL ON THE SAME TAB, it works as intended. It's just the first time loading this websites on a new tab/window.

**Comment by anishre** [comment]
I use Firefox 126 on Windows 10 and I also have the same issue. On Stable FF it doesn't work but on Nightly it does.
![image](https://github.com/darkreader/darkreader/assets/39543074/f24e440f-cc3c-45b1-a7d3-f797773f0cfb)
Dark Reader is disabled so I could see effect on both versions.

**Comment by Dziksonn** [comment]
Same problems here, if I disable and enable again, then it goes back to normal.
![obraz](https://github.com/user-attachments/assets/3b2ce662-93bc-42b7-a8a1-3a1d496830e6)

**Comment by jmcken1** [comment]
Same problem as everyone else. Happening on YouTube and GitHub – Dark Reader says “Dark theme detected”, but it defaults to “on” for these sites anyway, no matter _how many times_ I set them to “off”.

Weirdest part is that, at least for YouTube, this previously happened in the past but then stopped for a long time, with Dark Reader staying off. Then yesterday, I accidentally hit the hotkey to toggle DR on YouTube, and ever since that moment it’s been defaulting to on. No idea how hitting the hotkey messed that up.

**Comment by tigattack** [comment]
Adding affected domains here solved it for me: Dark Reader -> More -> All settings -> Site List.

**Comment by jmcken1** [comment]
> Adding affected domains here solved it for me: Dark Reader -> More -> All settings -> Site List.

Hey, that seems to have fixed it! Only tested real quick, but it’s working at the moment. Thanks!

**Comment by paulosouuza** [comment]
> Adding affected domains here solved it for me: Dark Reader -> More -> All settings -> Site List.

This fixed the issue for me too.

**Comment by NineKain** [comment]
Yeah an issue for me too

**Comment by Fxzzi** [comment]
having the exact same issue. has a fix been found yet?

**Comment by Myshor** [comment]
For now the only workaround is already given: https://github.com/darkreader/darkreader/issues/11990#issuecomment-2347334028

**Comment by chase-moskal** [comment]
i found a fix for my web app https://authlocal.org/ which aims to opt-out of dark-reader since it's already dark.

this fix is only relevant to confused developers like me who are trying to get darkreader to be disabled on their website -- i was doubly confused because, as this issue states, darkreader indicated "dark theme detected" and yet was still meddling

for developers like me who similarly land in this thread:

1. :x: for a long time, i was using this meta tag:
    ```html
    <meta name="darkreader" content="dark"/>
    ```
    - i believe this used to be working in the past
    - and then, some time ago, it stopped working
2. :white_check_mark: apparently nowadays darkreader expects this meta tag instead:
    ```html
    <meta name="darkreader-lock"/>
    ```
    - i found this documented [here in CONTRIBUTING.md](https://github.com/darkreader/darkreader/blob/main/CONTRIBUTING.md#disabling-dark-reader-on-your-site)

i've implemented this fix in my static site generator in [this commit](https://github.com/benevolent-games/turtle/commit/8a977eb093a82beccf2a566c0ea2cf5e8e661187#diff-7268a7676d56131d0d2c36e53bcb78faa578e2de15e9a4c2e50cb90f44b0aea1R20-R22) and now darkreader seems to be picking up this new meta tag and successfully disabling itself -- with this i have succeeded in my mission of making my apps play nice with darkreader.

however, i think **this is still a valid bug for end-users,** since when i was using the wrong/old meta tag -- darkreader would report "dark theme detected" and indicate that darkreader was disabled, and yet it proceeded to apply unwanted styles -- and i experienced the same as others reported, that when i'd flick darkreader to be enabled, nothing would happen, and then when flipping it back to disabled, the unwanted darkreader styles would vanish -- seemingly indicating some kind of bug in darkreader's initialization logic on the page... in other words, when dark reader initially detects a dark-theme website, it indicates it is disabled, and yet, is in fact, not disabled

**Comment by Pahasara** [comment]
I'm facing the same issue on firefox (v135) with dark reader (v4.9.100). Github, Youtube, ChatGPT and some other websites affected. Mostly occurs if I open a new tab within an already opened tab.

**Comment by SrS2225a** [comment]
I think what is most likely happening is that some kind of race condition is occurring. It is applying dark styling before/after an dark theme is detected. Something like that

**Comment by lumynou5** [comment]
@SrS2225a But it looks like sequential.

When the injected script detects builtin dark theme, it sends a message to the background script (search `MessageTypeCStoBG.DARK_THEME_DETECTED`), and the background script sends another message to the injected script, then the latter removes styles and stops the detector.

**Comment by alexanderby** [comment]
Does anyone have this issue in Chrome or Edge?

**Comment by SrS2225a** [comment]
@lumynou5 Yeah, that makes since. I was thinking an race condition since it is an hit or miss whenever dark reader detects the dark theme for some people like myself (like opening a new tab). My second thought was also thinking the same thing that it could be sequential. But to me the first option sounds more plausible

**Comment by SrS2225a** [comment]
@alexanderby People have been reporting this issue for all the browsers that dark reader supports, so yes

**Comment by alexanderby** [comment]
I cannot reproduce it in Chrome and I only see Firefox being mentioned in the comments.

**Comment by lumynou5** [comment]
> [@alexanderby](https://github.com/alexanderby) People have been reporting this issue for all the browsers that dark reader supports, so yes

Can you provide references?  I browsed through the comments on this issue, and found no browsers other than Firefox are mentioned.

**Comment by SrS2225a** [comment]
You're right—I went back through the comments and only see Firefox mentioned. I could've sworn I saw someone mentioning it for Chrome as well, but I must have been mistaken. My apologies for the oversight!

**Comment by alexanderby** [comment]
Finally it's done. I can't tell when it will be released as it takes Mozilla a week already to review a single character change from the previous release. So expect in 2-3 weeks.

**Comment by H3LLGHA5T** [comment]
I just enabled auto detection instead of enabling dark ready on every new website manually and saw that it didn't work on FF, glad it has been fixed now, I can wait a few weeks until the fix is live.

**Comment by mayuyoyuyo** [comment]
If you're using extensions on mobile it still defaults to on for everything for some reason. At least in chromium browsers, haven't tried firefox.

**Comment by bikram-agarwal** [comment]
I'm having this problem in Edge 146. Even if website has a dark theme, Dark Reader still overrides it with its own, even when I have "detect dark theme" enabled.

**Comment by Myshor** [comment]
> I'm having this problem in Edge 146. Even if website has a dark theme, Dark Reader still overrides it with its own, even when I have "detect dark theme" enabled.

Digging in something that was closed more than 6 months ago is dangerous. You will never know what you can discover after the digging. 😋

If you have the site which should be detected but it doesn't then the easiest way is to give some example(s).

I can write that I do not have any problems with all the sites I am using. Which you can not confirm as I am not giving any examples too. 😉

**Comment by lumynou5** [comment]
@bikram-agarwal Example? I can't reproduce.

Are you sure the website dark theme is detected? If it's a certain website you encounter problem, you should open a new issue to report broken website. This issue was for "dark theme detector works but dark reader doesn't auto-disable on **a random website**".
